### ISSUE TITLE
Test calling gd_errror(_ex) directly fails to build with automake

### ISSUE THREAD

**Issue by cmb69** [libgd/libgd#260]
I have written a test case for issue #252 which builds nicely with the cmake toolchain, but fails to link with the automake toolchain (default configuration). The relevant output:

    /bin/bash ../libtool  --tag=CC   --mode=link gcc -std=gnu99  -Werror -g -O2 -fvisibility=hidden -Wall   -o gd_error/bug00252 gd_error/bug00252.o libgdtest.a ../src/libgd.la -lm   -L/usr/lib/x86_64-linux-gnu -lz   -lpng12   -lfreetype   -lfontconfig -lfreetype    -ljpeg  -lXpm -lX11   -ltiff    -lwebp
    libtool: link: gcc -std=gnu99 -Werror -g -O2 -fvisibility=hidden -Wall -o gd_error/.libs/bug00252 gd_error/bug00252.o  libgdtest.a ../src/.libs/libgd.so -lm -L/usr/lib/x86_64-linux-gnu -lz -lpng12 -lfontconfig /usr/lib/x86_64-linux-gnu/libfreetype.so -ljpeg -lXpm -lX11 /usr/lib/x86_64-linux-gnu/libtiff.so -lwebp
    gd_error/bug00252.o: In function `main':
    /vagrant/cmb69-libgd/tests/gd_error/bug00252.c:16: undefined reference to `gd_error_ex'
    /vagrant/cmb69-libgd/tests/gd_error/bug00252.c:17: undefined reference to `gd_error'
    collect2: error: ld returned 1 exit status

`nm -g src/.libs/libgd.so | grep gd_error` produces no output.

Is there a bug in the automake toolchain, or am I doing something wrong?

**Comment by pierrejoye** [comment]
Do  you have a PR to look at? Maybe the link to libgd is missing?

**Comment by cmb69** [comment]
I've just submitted PR #261. If in doubt, we can wait for the travis checks. :-)

**Comment by pierrejoye** [comment]
I can see one problem, these functions are not exposed.

Anything from outside the library code (which is the case for a test case) should use:

typedef void(*gdErrorMethod)(int, const char *, va_list);

BGD_DECLARE(void) gdSetErrorMethod(gdErrorMethod);
BGD_DECLARE(void) gdClearErrorMethod(void);

By default these internal functions are used.

**Comment by cmb69** [comment]
I see, thanks! However, that does not solve the issue (nor does additionally BGD_DECLARE'ing gd_error and gd_error_ex in the test case). It seems to me that gd_error(_ex) would have to be exported when building libgd.so, but most likely not exporting these symbols is on purpose. So perhaps it's best to drop the test case.

Or we may consider to change the automake toolchain to behave like the cmake toolchain wrt. to tests. Apparently, Bin/libgd.so contains gd_error(_ex).

**Comment by pierrejoye** [comment]
They are not exposed on purposed.

We added in 2.1.x the private flags so non exposed APIs cannot be seen or
used.

**Comment by cmb69** [comment]
Okay, so I'll drop the test.

**Comment by pierrejoye** [comment]
Makes me wonder if we forgot to add them to cmake :)

**Comment by cmb69** [comment]
> Makes me wonder if we forgot to add them to cmake :)

Apparently not. `nm -g Bin/libgd.so | grep gd_error` reports for me:

    000000000000eae8 T gd_error
    000000000000eb8b T gd_error_ex

But these symbols are not contained anymore in the installed libgd.so.

**Comment by pierrejoye** [comment]
Exactly, we miss CMAKE_CXX_VISIBILITY_PRESET in CMake scripts.

However I am not too keen to use what they say here:

https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html

as it would required a special header for cmake to be included.

Alternatively adding:

    IF(CMAKE_COMPILER_IS_GNUCC)
         SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fvisibility=hidden")
    ENDIF(CMAKE_COMPILER_IS_GNUCC)

could do it too. Can you try? It should fail the same way than with autoconf.

**Comment by cmb69** [comment]
Appending these lines to CMakeLists.txt doesn't change anything. :?

**Comment by pierrejoye** [comment]
Are gd_error* still visible when you build with cmake? That's weird they should not if -fvisibility=hidden is passed

**Comment by cmb69** [comment]
> Are gd_error* still visible when you build with cmake?

Sorry, for the late reply. I have not fogotten about this issue, but rather concentrate on more urgent things instead. :-)

I'm coming back to this after 2.2.3 has been released.

**Comment by cmb69** [comment]
I finally got -fvisibility=hidden to take effect (maybe it didn't formerly because of a debug build). But that doesn't solve the issue.

I've found the culprit is that on my system HAVE_VISIBILITY is not properly set for the cmake toolchain. In config.h it's simply `/* #undef HAVE_VISIBILITY */`, and that leads to the [definition of empty macros](https://github.com/libgd/libgd/blob/master/src/gd.h#L72-L75).

With automake, HAVE_VISIBILITY is properly set, however.

It seems to me that this is bug in the cmake toolchain, but I don't know how to solve it.

**Comment by vapier** [comment]
> nm -g src/.libs/libgd.so | grep gd_error produces no output.

to clear up some things, `nm` only shows debugging information which means it's useless on stripped ELFs.  if you want to get useful results, look at `readelf -sW` instead.  symbols listed under `.dynsym` are the exported ABI while ones under `.symtab` are all the debugging symbols (which is usually a superset of `.dynsym`).  not that it makes a huge difference in this specific case.

> I finally got -fvisibility=hidden to take effect (maybe it didn't formerly because of a debug build).

the behavior of `-fvisibility=hidden` should behave the same regardless of `-O` and `-g` flags.  i'm not sure what you mean otherwise with "a debug build".

> However I am not too keen to use what they say here:
> https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html
> as it would required a special header for cmake to be included.

that's not specific to cmake.  we've just been playing fast & loose with our normal builds :).  we can't use any defines from `config.h` in `gd.h` (or any other exported header) unless we take special care.  which we haven't.

> With automake, HAVE_VISIBILITY is properly set, however.

well, it's set at build time, but not at installed time (when other packages include the header).

gcc added support for the `visibility` attribute starting with the 3.3 release which was over [13 years ago](https://gcc.gnu.org/gcc-3.3/).  if we don't care about older versions of gcc, we can just delete the `HAVE_VISIBILITY` define entirely and always use the `__attribute__((visibility(...)))` logic.  if we do care about older gcc versions, we could add a `__GNUC__` version test.

**Comment by vapier** [comment]
also along the `readelf` lines, the output is a bit more verbose and useful.  for example, `gd_error` looks like:
```
   222: 000000000000ea00   186 FUNC    LOCAL  DEFAULT   11 gd_error_ex
   290: 000000000000e930   199 FUNC    LOCAL  DEFAULT   11 gd_error
```

we can see that it's not being exported.

**Comment by vapier** [comment]
what do you think of commit b22313b5cdeeadffdcef2520c8b282d32338a663 ?  it works for me -- the exported symbols of cmake's libgd.so matches that of autotool's libgd.so.

it also cleans up the `HAVE_VISIBILITY` parts and doesn't have the downsides you mentioned in that export header macro.

**Comment by cmb69** [comment]
Thanks for the explanations regarding `nm` vs. `readelf` – quite helpful!

> i'm not sure what you mean otherwise with "a debug build".

Originally, I had appended Pierre's suggestion to the end of CMakeLists.txt, and that didn't seem to have any effect. Not sure why, but it doesn't matter. :-)

> what do you think of commit b22313b?

Works for me, too. And I agree that supporting GCC < 3.3 shouldn't be necessary anymore.

**Comment by vapier** [comment]
ok, i've pushed that fix to master & the 2.2 branch.  i think that's all we needed here.
